desc/integrals/singularities.py

```diff
@@ -433,91 +558,109 @@ def eval_pt_loop(i, fj):
-        # omega and add it back in
-        source_data_polar["zeta"] = zeta_i
-        source_data_polar["theta"] = theta_i
-        if "phi" in keys:
-            source_data_polar["phi"] = zeta_i + source_data_polar["omega"]
+        # The polar node coordinates should be shifts around the evaluation point.
+        # That is where the polar basis vectors need to be computed.
+        # These coordinates should also be the (θ, ζ) coordinates at which the maps
+        # above were interpolated. For FFT interpolator however, the maps were
+        # interpolated to shifts around the source point. When NFP is different
+        # these are not the same points. Since the components of maps in coordinates
+        # of the orthonormal polar basis are the same, we can pretend they were
+        # interpolated to polar nodes around the evaluation point.
+        source_data_polar["theta"] = eval_theta + interpolator.shift_t[i]
+        source_data_polar["zeta"] = eval_zeta + interpolator.shift_z[i]
+        # ϕ is not periodic map of θ, ζ.
+        if "omega" in keys:
+            # TODO (#465): For nonzero ω, the quadrature may not be symmetric about the
+            #  singular point for hypersingular kernels such as the Biot-Savart
+            #  kernel. (Recall the singularity is in real space). Hence the quadrature
+            #  may not converge to the desired Hadamard finite part. Prove otherwise or
+            #  use uniform grid in θ, ϕ and map coordinates before starting the singular
+            #  integral routine.
+            source_data_polar["phi"] = (
+                source_data_polar["zeta"] + source_data_polar["omega"]
+            )
 
         # eval pts x source pts for 1 polar grid offset
         # only need diagonal term because polar grid points
-        # don't contribute to other eval pts
-        k = kernel(
-            {key: val for key, val in eval_data.items() if key in keys},
-            source_data_polar,
-            diag=True,
-        ).reshape((eval_grid.num_nodes, kernel.ndim))
-
-        dS = (v[i] * source_data_polar["|e_theta x e_zeta|"])[:, None]
-        fi = k * dS
+        # don't contribute to other eval pts.
+        k = kernel(eval_data, source_data_polar, diag=True).reshape(
+            eval_grid.num_nodes, kernel.ndim
+        )
+        dS = v[i] * source_data_polar["|e_theta x e_zeta|"]
+        fi = k * dS[:, jnp.newaxis]
         return fi
 
     def polar_pt_loop(i, f):
-        # this calculates the effect at a single evaluation point, from all others
-        # in a single field period. loop this to get all pts
-        f_temp = polar_pt_vmap(i)
-        return f + f_temp.reshape((eval_grid.num_nodes, kernel.ndim))
+        return f + polar_pt(i)
 
-    f = jnp.zeros((eval_grid.num_nodes, kernel.ndim))
     # vmap found more efficient than fori_loop, esp on gpu, but uses more memory
     if loop:
-        f = fori_loop(0, v.size, polar_pt_loop, f)
+        f = fori_loop(
```

Comment:
i think also worth using vmap_chunked or batch_map or whatever here

Reply:
This is a case where chunking does not generalize the for loop because the loop reduces as the sum is computed (so max array size is 1), while chunking does not.

Should I heed your old code comment and interpret chunk size=1 as `fori_loop` and `chunk_size!=1` as a chunked vmap? Or just the latter?

Reply:
ahh good point. Then yeah I think leave the loop for chunksize=1 but use vmap_chunked for the else

Reply:
not for this PR, but at some point might be worth implementing some sort of batched reduction `desc.batching`